### PR TITLE
CSHARP-3669: Support loadBalanced URI option (v2).

### PR DESCRIPTION
**Pull request by DmitryLukyanov**

### CODE REVIEW

**Comment by DmitryLukyanov** [comment]
Added @MikalaiMazurenka , because he was reviewer for the original PR

File: src/MongoDB.Driver/MongoClientSettings.cs [hunk 1]
```diff
@@ -879,6 +894,7 @@ public static MongoClientSettings FromUrl(MongoUrl url)
             clientSettings.ReplicaSetName = url.ReplicaSetName;
             clientSettings.RetryReads = url.RetryReads.GetValueOrDefault(true);
             clientSettings.RetryWrites = url.RetryWrites.GetValueOrDefault(true);
+            clientSettings.LoadBalanced = url.LoadBalanced;
```
Comment (rstam) [inline src/MongoDB.Driver/MongoClientSettings.cs:897]: Put in alphabetical order (and move `LocalThreshold` also).

File: src/MongoDB.Driver/MongoClientSettings.cs [hunk 2]
```diff
@@ -931,6 +947,7 @@ public MongoClientSettings Clone()
             clone._replicaSetName = _replicaSetName;
             clone._retryReads = _retryReads;
             clone._retryWrites = _retryWrites;
+            clone._loadBalanced = _loadBalanced;
```
Comment (rstam) [inline src/MongoDB.Driver/MongoClientSettings.cs:950]: Put in alphabetical order (and move `LocalThreshold` also).

File: src/MongoDB.Driver/MongoClientSettings.cs [hunk 3]
```diff
@@ -985,6 +1002,7 @@ public override bool Equals(object obj)
                 _heartbeatInterval == rhs._heartbeatInterval &&
                 _heartbeatTimeout == rhs._heartbeatTimeout &&
                 _ipv6 == rhs._ipv6 &&
+                _loadBalanced == rhs._loadBalanced &&
```
Comment (rstam) [inline src/MongoDB.Driver/MongoClientSettings.cs:1005]: Move `localThreshold` here also to preserve alphabetical order.

File: src/MongoDB.Driver/MongoClientSettings.cs [hunk 4]
```diff
@@ -1067,6 +1085,7 @@ public override int GetHashCode()
                 .Hash(_heartbeatInterval)
                 .Hash(_heartbeatTimeout)
                 .Hash(_ipv6)
+                .Hash(_loadBalanced)
```
Comment (rstam) [inline src/MongoDB.Driver/MongoClientSettings.cs:1088]: Move localThreshold here also to preserve alphabetical order.

File: src/MongoDB.Driver/MongoClientSettings.cs [hunk 5]
```diff
@@ -1131,6 +1150,10 @@ public override string ToString()
             sb.AppendFormat("HeartbeatInterval={0};", _heartbeatInterval);
             sb.AppendFormat("HeartbeatTimeout={0};", _heartbeatTimeout);
             sb.AppendFormat("IPv6={0};", _ipv6);
+            if (_loadBalanced)
+            {
+                sb.AppendFormat("LoadBalanced={0};", _loadBalanced);
+            }
```
Comment (rstam) [inline src/MongoDB.Driver/MongoClientSettings.cs:1156]: Move localThreshold here also to preserve alphabetical order.

File: src/MongoDB.Driver/MongoUrlBuilder.cs [hunk 6]
```diff
@@ -120,6 +121,7 @@ public MongoUrlBuilder()
             _replicaSetName = null;
             _retryReads = null;
             _retryWrites = null;
+            _loadBalanced = false;
```
Comment (rstam) [inline src/MongoDB.Driver/MongoUrlBuilder.cs:124]: Keep in alphabetical order (move `localThreshold` also).

File: src/MongoDB.Driver/MongoUrlBuilder.cs [hunk 7]
```diff
@@ -833,6 +847,7 @@ public void Parse(string url)
             _replicaSetName = connectionString.ReplicaSet;
             _retryReads = connectionString.RetryReads;
             _retryWrites = connectionString.RetryWrites;
+            _loadBalanced = connectionString.LoadBalanced;
```
Comment (rstam) [inline src/MongoDB.Driver/MongoUrlBuilder.cs:850]: Keep in alphabetical order (move localThreshold also).

File: src/MongoDB.Driver/MongoUrlBuilder.cs [hunk 8]
```diff
@@ -1078,6 +1093,10 @@ public override string ToString()
             {
                 query.AppendFormat("minPoolSize={0};", _minConnectionPoolSize);
             }
+            if (_loadBalanced)
+            {
+                query.AppendFormat("loadBalanced={0};", JsonConvert.ToString(_loadBalanced));
+            }
```
Comment (rstam) [inline src/MongoDB.Driver/MongoUrlBuilder.cs:1099]: Keep in alphabetical order (move localThreshold also).

File: tests/MongoDB.Driver.Tests/MongoClientSettingsTests.cs [hunk 9]
```diff
@@ -543,30 +557,53 @@ public void TestFreeze()
...
+            {
+                settings.LoadBalanced = true;
+                settings.ReplicaSetName = "test";
+            });
 
-            settings = new MongoClientSettings();
-            settings.DirectConnection = true;
-            var endpoint = "test5.test.build.10gen.cc:53";
-            settings.Servers = new[] { MongoServerAddress.Parse(endpoint), MongoServerAddress.Parse(endpoint) };
+            void AssertOptions(Action<MongoClientSettings> setAction)
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/MongoClientSettingsTests.cs:598]: The name of this method confused me.

Can we name it `AssertException` instead?

**Review by rstam — CHANGES_REQUESTED** [review]
Looks good with some minor changes requested.

File: src/MongoDB.Driver/MongoClientSettings.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoClientSettings.cs]: done

File: src/MongoDB.Driver/MongoClientSettings.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoClientSettings.cs]: done

File: src/MongoDB.Driver/MongoClientSettings.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoClientSettings.cs:1005]: done

File: src/MongoDB.Driver/MongoClientSettings.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoClientSettings.cs:1088]: done

File: src/MongoDB.Driver/MongoClientSettings.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoClientSettings.cs:1156]: done

File: src/MongoDB.Driver/MongoUrlBuilder.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoUrlBuilder.cs]: done

File: src/MongoDB.Driver/MongoUrlBuilder.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoUrlBuilder.cs]: done

File: src/MongoDB.Driver/MongoUrlBuilder.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoUrlBuilder.cs]: done

File: tests/MongoDB.Driver.Tests/MongoClientSettingsTests.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/MongoClientSettingsTests.cs]: done

**Review by DmitryLukyanov — COMMENTED** [review]
ready for review

**Review by rstam — APPROVED** [review]
LGTM

**Pull request merged**